### PR TITLE
Add QEMU logging to PyPANDA

### PR DESCRIPTION
**Pull request by AndrewFasano**
Allows for dynamically changing the log level from pypanda. Useful for enabling things like in_asm logging for specific parts of an execution.

Could be made better with some helper functions which include the mappings between log level names and their integer values.

### CODE REVIEW

File: panda/python/core/pandare/qemu_logging.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
+'''
+QEMU_Log_Manager
```
Comment (jamcleod) [inline panda/python/core/pandare/qemu_logging.py]: Documentation should probably include a list of log event types

File: panda/python/core/pandare/qemu_logging.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            if self.panda.libpanda.fileno(self.panda.libpanda.qemu_logfile) != 2:
+                self.panda.libpanda.fclose(self.panda.libpanda.qemu_logfile)
+            self.panda.libpanda.qemu_logfile = self.panda.ffi.NULL
+    
+    def output_to_stdout(self):
+        """Removes the current log file and outputs to stdout"""
+        self.remove_log_file()
+
+
+
```
Comment (jamcleod) [inline panda/python/core/pandare/qemu_logging.py]: - Trailing newlines
- Double checking: Is `qemu_logfile == NULL` enough to output to stdout?

File: panda/python/core/pandare/qemu_logging.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        }
+        self.log_members = log_members
+
+        def setup_member(member):
+            def enable():
+                self.panda.libpanda.qemu_loglevel |= (1 << log_members[member])
+            def disable():
+                self.panda.libpanda.qemu_loglevel &= ~(1 << log_members[member])
+            name = member.lower()
+            setattr(self, name + "_enable", enable)
```
Comment (jamcleod) [inline panda/python/core/pandare/qemu_logging.py]: would doing this via a metaclass allow this to show up in pdoc? haven't actually tried before

**Review by jamcleod — CHANGES_REQUESTED** [review]
Some minor nitpicks regarding documentation, the only thing I think is strictly necessary is adding a list of log event types to the docs

File: panda/python/core/pandare/qemu_logging.py (hunk 1 above)
Comment (lacraig2) [reply panda/python/core/pandare/qemu_logging.py]: Good idea. Added.

File: panda/python/core/pandare/qemu_logging.py (hunk 2 above)
Comment (lacraig2) [reply panda/python/core/pandare/qemu_logging.py]: - Good catch
- It did work, but I didn't like how so I did something of a rewrite.

File: panda/python/core/pandare/qemu_logging.py (hunk 3 above)
Comment (lacraig2) [reply panda/python/core/pandare/qemu_logging.py]: Nope. In the last commit I actually removed that part. It's not very QEMU-esque anyhow.

**Comment by lacraig2** [comment]
This PR does more now. It does the following:
- Dynamically changing logging level
- Allows logging file open/close dynamically
- Allows logging file redirection to stderr/stdout dynamically
- Creates new log for `PANDA` usage (not the pandalog and not the same as the RR log)
- Exposes `log` function that outputs to qemu_log for the `PANDA` trace.

**Comment by lacraig2** [comment]
@jamcleod you did so well with your review that I rethought a lot of it.

**Comment by AndrewFasano** [comment]
This is looking really cool. Thanks for making this PR so much better

**Review by jamcleod — APPROVED** [review]

**Comment by AndrewFasano** [comment]
I think this PR broke dev: create_panda_datatypes.py is now failing because `qlog.h` is missing.

Hopefully @lacraig2 still has that file somewhere and there's an easy fix. But this raises a bigger issue which is that (apparently) changes to create_panda_datatypes.h aren't being tested by CI?

**Comment by AndrewFasano** [comment]
I just restored the branch and it looks like CI was failing on the last commit, even though it seems like it passed for the merge commit. I'm not sure what happened?

In commit https://github.com/panda-re/panda/commit/5327af37ce321358b7af7f31c93884752332bebd the CI tests failed with failures such as https://github.com/panda-re/panda/runs/4809633886?check_suite_focus=true and https://github.com/panda-re/panda/runs/4809706280?check_suite_focus=true

**Comment by AndrewFasano** [comment]
I'm seeing an occasional crash with this after I set up a log file and turn on logging, not a big deal but thought I'd provide the trace

```
@blocking
def some_func():
  panda.qlog.output_to_file("log.txt")
  panda.qlog.enable("TB_IN_ASM")
  panda.run_serial_cmd("whoami")
```

```
___fprintf_chk (fp=fp@entry=0x0, flag=flag@entry=1, format=format@entry=0x7ffff1beebb7 "0x%08x:  ") at fprintf_chk.c:30
30      fprintf_chk.c: No such file or directory.
(gdb) bt
#0  0x00007ffff7b142bb in ___fprintf_chk (fp=fp@entry=0x0, flag=flag@entry=1, format=format@entry=0x7ffff1beebb7 "0x%08x:  ") at fprintf_chk.c:30
#1  0x00007ffff1814bd8 in fprintf (__fmt=0x7ffff1beebb7 "0x%08x:  ", __stream=0x0) at /usr/include/x86_64-linux-gnu/bits/stdio2.h:97
#2  0x00007ffff1814bd8 in target_disas (out=0x0, cpu=cpu@entry=0x12c1b20, code=code@entry=3221373840, size=20, flags=<optimized out>) at /home/fasano/git/panda/disas.c:406
#3  0x00007ffff190959b in log_target_disas (flags=<optimized out>, len=<optimized out>, start=3221373840, cpu=0x12c1b20) at /home/fasano/git/panda/include/exec/log.h:43
#4  0x00007ffff190959b in gen_intermediate_code (env=env@entry=0x12c9df8, tb=tb@entry=0x7fffd0503010) at /home/fasano/git/panda/target/arm/translate.c:12129
#5  0x00007ffff17efd79 in tb_gen_code (cpu=cpu@entry=0x12c1b20, pc=pc@entry=3221373840, cs_base=cs_base@entry=0, flags=flags@entry=268959744, cflags=<optimized out>, cflags@entry=0)
    at /home/fasano/git/panda/translate-all.c:1352
#6  0x00007ffff17f4e04 in tb_find (tb_exit=0, last_tb=<optimized out>, cpu=0x0) at /home/fasano/git/panda/cpu-exec.c:427
#7  0x00007ffff17f4e04 in cpu_exec (cpu=cpu@entry=0x12c1b20) at /home/fasano/git/panda/cpu-exec.c:836
#8  0x00007ffff18170c0 in tcg_cpu_exec (cpu=0x12c1b20) at /home/fasano/git/panda/cpus.c:1200
#9  0x00007ffff18170c0 in qemu_tcg_cpu_thread_fn (arg=<optimized out>) at /home/fasano/git/panda/cpus.c:1298
#10 0x00007ffff77ca6db in start_thread (arg=0x7fffcaff6700) at pthread_create.c:463
#11 0x00007ffff7b0371f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) up 3
#3  0x00007ffff190959b in log_target_disas (flags=<optimized out>, len=<optimized out>, start=3221373840, cpu=0x12c1b20) at /home/fasano/git/panda/include/exec/log.h:43
43          target_disas(qemu_logfile, cpu, start, len, flags);
(gdb) p qemu_logfile
$1 = (FILE *) 0x0
```

**Pull request merged**